R/utils.R

```diff
@@ -259,82 +259,500 @@ rowcol_to_dim <- function(row, col) {
+#' dims_column_names <- wb_dims(x = mtcars, select = "col_names")
+#' wb$add_font(dims = dims_column_names, bold = TRUE, size = 13)
+#'
+#' # Finally, to add styling to column "cyl" (the 4th column) (only the data)
+#' # there are many options, but here is the preferred one
+#' # if you know the column index, wb_dims(x = mtcars, cols = 4) also works.
+#' dims_cyl <- wb_dims(x = mtcars, cols = "cyl")
+#' wb$add_fill(dims = dims_cyl, color = wb_color("pink"))
+#'
+#' # Mark a full column as important(with the column name too)
+#' wb_dims_vs <- wb_dims(x = mtcars, cols = "vs", select = "x")
+#' wb$add_fill(dims = wb_dims_vs, fill = wb_color("yellow"))
+#' wb$add_conditional_formatting(dims = wb_dims(x = mtcars, cols = "mpg"), type = "dataBar")
+#' # wb_open(wb)
+wb_dims <- function(..., select = NULL) {
   args <- list(...)
-  nams <- names(args)
+  lengt <- length(args)
 
-  col_names <- args$col_names
-  row_names <- args$row_names
+  if (lengt == 0 || (lengt == 1 && is.null(args[[1]]))) {
+    return("A1")
+  }
 
-  cnam_null <- is.null(col_names)
-  rnam_null <- is.null(row_names)
 
-  srow <- args$start_row
-  scol <- args$start_col
+  # nams cannot be NULL now
+  nams <- names(args) %||% rep("", lengt)
+  valid_arg_nams <- c("x", "rows", "cols", "from_row", "from_col", "row_names", "col_names")
+  any_args_named <- any(nzchar(nams))
+  # unused, but can be used, if we need to check if any, but not all
+  # Check if valid args were provided if any argument is named.
+  if (any_args_named) {
+    if (any(c("start_col", "start_row") %in% nams)) {
+      stop("Use `from_row` / `from_col` instead of `start_row` / `start_col`")
+    }
+    match.arg_wrapper(arg = nams, choices = c(valid_arg_nams, ""), several.ok = TRUE, fn_name = "`wb_dims()`")
+  }
+  # After this point, no need to search for invalid arguments!
 
-  scol_null <- is.null(scol)
-  srow_null <- is.null(srow)
+  n_unnamed_args <- length(which(!nzchar(nams)))
+  all_args_unnamed <- n_unnamed_args == lengt
+  # argument dispatch / creation here.
+  # All names provided, happy :)
+  # Checking if valid names were provided.
 
-  if (srow_null) srow <- 0 else srow <- srow - 1L
-  if (scol_null) scol <- 0 else scol <- col2int(scol) - 1L
+  if (n_unnamed_args > 2) {
+    stop("only `rows` and `cols` can be provided without names. You must name all other arguments.")
+  }
+  if (lengt == 1 && all_args_unnamed) {
+    stop(
+      "Supplying a single unnamed argument is not handled by `wb_dims()`",
+      "use `x`, `from_row` / `from_col`. You can also use `dims = NULL`"
+    )
+  }
+  ok_if_arg1_unnamed <-
+    is.atomic(args[[1]]) || any(nams %in% c("rows", "cols"))
 
-  x <- args[[1]]
-  exp_name <- inherits(x, "data.frame") || inherits(x, "matrix")
+  if (nams[1] == "" && !ok_if_arg1_unnamed) {
+    stop(
+      "The first argument must either be named or be a vector.",
+      "Providing a single named argument must either be `from_row`, `from_col` or `x`."
+    )
+  }
```

Comment:
Please add empty lines after a closing `}`. This already is a huge function, these blank lines will not hurt us